eppo_core/src/ufc/compiled_flag_config.rs

```diff
@@ -0,0 +1,265 @@
+    bandits: HashMap<String, Vec<BanditVariationWire>>,
+) -> HashMap<String, HashMap<String, BanditVariationWire>> {
+    bandits
+        .into_iter()
+        .flat_map(|(_, bandits)| bandits.into_iter())
+        .fold(HashMap::new(), |mut acc, variation| {
+            acc.entry(variation.flag_key.clone())
+                .or_default()
+                .insert(variation.variation_value.clone(), variation);
+            acc
+        })
+}
+
+fn compile_flag(meta_data: EventMetaData, flag: FlagWire) -> Flag {
+    let variation_values = flag
+        .variations
+        .into_values()
+        .map(|variation| {
+            let assignment_value = variation
+                .value
+                .into_assignment_value(flag.variation_type)
+                .ok_or(EvaluationFailure::Error(
+                    EvaluationError::UnexpectedConfigurationError,
+                ));
+
+            (variation.key, assignment_value)
+        })
+        .collect::<HashMap<_, _>>();
+
+    let allocations = flag
+        .allocations
+        .into_iter()
+        .map(|allocation| {
+            compile_allocation(
+                meta_data,
+                &flag.key,
+                allocation,
+                &variation_values,
+                flag.total_shards,
+            )
+        })
+        .collect();
+
+    Flag {
+        variation_type: flag.variation_type,
+        allocations,
+    }
+}
+
+fn compile_allocation(
+    meta_data: EventMetaData,
+    flag_key: &Str,
+    allocation: AllocationWire,
+    variation_values: &HashMap<Str, Result<AssignmentValue, EvaluationFailure>>,
+    total_shards: u32,
+) -> Allocation {
+    let splits = allocation
+        .splits
+        .into_iter()
+        .map(|split| {
+            compile_split(
+                meta_data,
+                flag_key,
+                &allocation.key,
+                split,
+                variation_values,
+                total_shards,
+                allocation.do_log,
+            )
+        })
+        .collect();
+    Allocation {
+        key: allocation.key,
+        start_at: allocation.start_at,
+        end_at: allocation.end_at,
+        rules: allocation.rules,
+        splits,
+    }
+}
+
+fn compile_split(
+    meta_data: EventMetaData,
+    flag_key: &Str,
+    allocation_key: &Str,
+    split: SplitWire,
+    variation_values: &HashMap<Str, Result<AssignmentValue, EvaluationFailure>>,
+    total_shards: u32,
+    do_log: bool,
+) -> Split {
+    let shards = split
+        .shards
+        .into_iter()
+        .filter_map(|shard|
+                            // `compile_shard` may return `None` for shards that are
+                            // "insignificant", meaning that they *always* match, so they don't even
+                            // need to be checked. We filter out such shards here with
+                            // `.filter_map()`.
+                            compile_shard(shard, total_shards))
```

Comment:
📓 filtering out insignificant shards

Reply:
More explanation.

A split is selected when the subject matches all shards in the split. i.e., to match a split, the subject must satisfy *all* shards; if there're no shards in the split, that's an automatic match.

So we have the following formula:
```
matches_split = true && matches_shard1 && matches_shard2 && ... && matches_shardN
```

A shard that contains a $[0..total\\_shards)$ range always evaluates to `true` because the shard value will always fall in that range ($\forall hash \in \mathbb{N}: 0 \le hash\ \\%\ total\\_shards < total\\_shards$), and that range will dominate the shard (because shard matches if any range matches (OR), and $\forall x: x \lor true = true$).

So we can replace always-matching shard with a `true` in the original equation:

```
matches_split = true && matches_shard1 && true && ... && matches_shardN
```

And because we know that `true` is an identity for `&&` ($\forall x: x \land true = x$), we can drop `true`s from the formula. So we can drop that shard altogether:

```
matches_split = matches_shard1 && ... && matches_shardN
```